replace `#[allow_internal_unstable]` with `#[rustc_allow_const_fn_unstable]` for `const fn`s

liketechnik opened the pull request:
`#[allow_internal_unstable]` is currently used to side-step feature gate and stability checks.
While it was originally only meant to be used only on macros, its use was expanded to `const fn`s.

This pr adds stricter checks for the usage of `#[allow_internal_unstable]` (only on macros) and introduces the `#[rustc_allow_const_fn_unstable]` attribute for usage on `const fn`s.

This pr does not change any of the functionality associated with the use of `#[allow_internal_unstable]` on macros or the usage of `#[rustc_allow_const_fn_unstable]` (instead of `#[allow_internal_unstable]`) on `const fn`s (see https://github.com/rust-lang/rust/issues/69399#issuecomment-712911540).

Note: The check for `#[rustc_allow_const_fn_unstable]` currently only validates that the attribute is used on a function, because I don't know how I would check if the function is a `const fn` at the place of the check. I therefore openend this as a 'draft pull request'.

Closes rust-lang/rust#69399

r? @oli-obk

liketechnik committed:
add rustc_allow_const_fn_unstable attribute

allow_internal_unstable is currently used
to side-step feature gate and stability checks.
While it was originally only meant to be used
only on macros, its use was expanded to
const functions.

This commit prepares stricter checks for the usage of allow_internal_unstable (only on macros)
and introduces the rustc_allow_const_fn_unstable attribute for usage on functions.

See rust-lang/rust#69399

liketechnik committed:
switch allow_internal_unstable const fns to rustc_allow_const_fn_unstable

liketechnik committed:
validate allow_internal_unstable target

Adds a check to make sure `#[allow_internal_unstable]`
can be applied only to macro definitions.

liketechnik committed:
validate rustc_allow_const_fn_unstable targets

Adds a check to make sure `#[rustc_allow_const_fn_unstable]`
can be applied only to functions.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @oli-obk (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

oli-obk commented on compiler/rustc_expand/src/base.rs:
```diff
@@ -768,8 +768,13 @@ impl SyntaxExtension {
+        let allow_internal_unstable = {
+            let mut feat_list = Vec::new();
+            attr::allow_internal_unstable(sess, &attrs).map(|features| feat_list.extend(features));
+            attr::rustc_allow_const_fn_unstable(sess, &attrs)
```
I don't think we need this here. we just want `allow_internal_unstable` here, it's about macro expansion after all.

oli-obk commented on compiler/rustc_mir/src/transform/check_consts/mod.rs:
```diff
@@ -79,7 +79,7 @@ pub fn is_lang_panic_fn(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
 
 pub fn allow_internal_unstable(tcx: TyCtxt<'tcx>, def_id: DefId, feature_gate: Symbol) -> bool {
```
please also rename this function

oli-obk commented on compiler/rustc_passes/src/check_attr.rs:
```diff
@@ -791,6 +793,26 @@ impl CheckAttrVisitor<'tcx> {
+        target: Target,
+    ) -> bool {
+        if let Target::Fn | Target::Method(_) = target {
+            // FIXME Check that this isn't just a function, but a const fn
```
you an use `self.tcx.is_const_fn_raw(self.tcx.hir().local_def_id(hir_id))` to check whether the function should be treated as a const fn.

oli-obk commented:
This is really clean, thanks for taking the care to build clear individual commits

liketechnik committed:
ignore #[rustc_allow_const_fn_unstable] for macro expansion

Recognition for `rustc_allow_const_fn_unstable` attribute was errorneously
added in 05f4a9a42abbbdb64d3ec4ed0ae6883c10c66e3f.

liketechnik committed:
rename allow_internal_unstable() to rustc_allow_const_fn_unstable() in rustc_mir

Followup rename from 05f4a9a42abbbdb64d3ec4ed0ae6883c10c66e3f,
which introduced `#[rustc_allow_const_fn_unstable]` for `const fn`s.

liketechnik committed:
fix validation for rustc_allow_const_fn_unstable targets

The validation was introduced in 3a63bf02998f7b5e040a4b87e049d03ddd144f74
without strict validation of functions, e. g. all function types were
allowed.
Now the validation only allows `const fn`s.

liketechnik commented:
Note: If I understood the contributing guide correctly, it is discouraged to change already pushed commits after review (e. g. interactive rebase). Instead it is preferred to add new commits containing the necessary changes, which I did now. If I got that wrong, I'm happy to rebase instead.

oli-obk commented:
The rule is there to make reviewing easier (since in large PRs it can quickly become infeasible to review the entire PR again), but once everything is approved fixing up history is definitely ok.

So... I'd r+ this as it is, though if you want to fix up history, that's totally alright with me, but not neccessary. Please let me know what you prefer

liketechnik commented:
I'd leave it as is this time, because the fix-up commits should be descriptive enough on their own.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit ac2c599f23d6c8b5d3413febde4c6febff68062e has been approved by `oli-obk`

<!-- @bors r=oli-obk ac2c599f23d6c8b5d3413febde4c6febff68062e -->
<!-- homu: {"type":"Approved","sha":"ac2c599f23d6c8b5d3413febde4c6febff68062e","approver":"oli-obk"} -->
